File: combinations.py
# LeetCode 77. Combinations Python
# dfs;

class Solution:
    def combine(self, n: int, k: int) -> list[list[int]]:
        ans = []
        self.count = 0

        def dfs(start, nums):
            if self.count == k:
                ans.append(nums)
                return

            for i in range(start, n+1):
                self.count += 1
                # print('Going in:', i+1, nums+[i])
                dfs(i+1, nums+[i])
                # print('Out with: ', i, nums)
                self.count -= 1

        dfs(1, [])

        return ans

n = 7
k = 3

sol = Solution()
ans = sol.combine(n, k )
print(ans)
        

File: combinations.php
<?php
/**
 * LeetCode 77. Combinations PHP DFS
*/
class Solution {
    /**
     * @param Integer $n
     * @param Integer $k
     * @return Integer[][]
     */
    function combine($n, $k) {
        $this->ans = [];
        $this->count = 0;
        $this->n = $n;
        $this->k = $k;

        $this->dfs(1, []);

        return $this->ans;
    }

    function dfs($start, $comb)
    {
        if($this->count == $this->k ){
            $this->ans[] = $comb;
            return;
        }

        if( $start > $this->n ) return;
        foreach( range( $start, $this->n ) as $i ){
            $this->count++;
            $this->dfs($i+1, array_merge($comb, [$i]));
            $this->count--;
        }
    }
}

$n = 4;
$k = 2;

$sol = new Solution();
$ans = $sol->combine($n, $k );
print_r($ans);
        